### ISSUE TITLE
How to access graph token from a backend with credentials (or something similar) of the current sharepoint loggedin user?

### ISSUE THREAD

**Issue by RafatRashid** [SharePoint/sp-dev-docs#6372]
## Category

- [x] Question
- [ ] Typo
- [ ] Additional article idea

## Question
Hello
I am hoping to access graph content endpoint for converting and downloading a file to pdf using an spfx context menu extension. **The clients are not willing to add permissions in sharepoint which is requested in webApiRequest**. Also **wont give an AD client ID with application permission**. So what we are trying to do is get a delegated client id and use delegated permissions to get the file from an azure function backend without invoking any login flow. However when I get the try to access the resource using graph client for C# the resource sends the following error:

"error": {
        "code": "AccessDenied",
        "message": "Either scp or roles claim need to be present in the token.",
        "innerError": {
            "date": "2020-10-13T05:47:46",
            "request-id": "d1dad236-c3f3-45ab-a243-297e7bf71047",
            "client-request-id": "d1dad236-c3f3-45ab-a243-297e7bf71047"
        }
    }

I was thinking of getting a token in SPFX environment and pass it to the function for requesting. But I am getting :

Uncaught (in promise) Error: interaction_required: AADSTS65001: The user or administrator has not consented to use the application with ID '675db3fc-53db-4b32-86ff-cc29b9d8b743' named 'SharePoint Online Client Extensibility Web Application Principal'. Send an interactive authorization request for this user and resource.
Trace ID: 92498e3a-1978-412e-9f83-86f623d31100
Correlation ID: 9729cd9c-94d8-4ecf-afa5-c68120b1760b
Timestamp: 2020-10-13 06:12:32Z

while trying to get the token using aadTokenProviderFactory.

I am not sure if my approach is ok. Can anyone please point me how can I access the resource with delegated permission maintaining the client requirements? 

## Environment details (*development & target environment*)
- **Your Developer Environment**: Windows 10
- **Target Environment**: SharePoint Online
- **Framework**: .NET Core v2.0 | Node.js v10
- **Tooling**: [VS Code | SPFx v1.10.0 | Visual Studio 2019]

**Comment by andrewconnell** [comment]
In order to query Microsoft Graph, you must obtain an access token. To get this, you need an Azure AD app that has been granted the necessary permissions on Microsoft Graph. There's _no_ way around this; that's how it works.

In your first scenario, it sounds like you're saying "we can't use the SharePoint Online Azure AD app because we cant grant it permissions" and "won't create Azure AD app and grant it permissions". So this won't work because of what I said above.

Your second scenario with SPFx is going to run into the exact same problem because it uses the SharePoint Online Azure AD app to obtain it's access token. If you can't grant it permission to Microsoft Graph, you can't use the **AadTokenProviderFactory** API. It requires the endpoint you're going to call and uses that to ask Azure AD for an access token for that endpoint that the SPO AAD app has been granted to use.

**Comment by RafatRashid** [comment]
Yes I undesrstand what you mean. I have a client id with delegated permission scoped for files read write and user read. Is there any way to utilize this client from either spfx extension or c# backend to access a graph endpoint? 
I am trying to access the content endpoint for getting pdf of doc files.

**Comment by andrewconnell** [comment]
@RafatRashid said:
> Is there any way to utilize this client from either spfx extension
Sort of... assuming it's a REST based HTTP endpoint, you can call it from SPFx. You can use the **HttpClient** or **AadHttpClient** (if it's secured) to call your endpoint from SPFx. Then your code in the HTTP endpoint can use your Azure AD app to obtain a token that you can use to call MS Graph.

> or c# backend to access a graph endpoint?
Yes... that's exactly how this is all designed to work. the Graph docs have tons of examples that show how to do this, but that's outside the scope of this repo which is specific to SP dev topics.

**Comment by RafatRashid** [comment]
Is there any way where I can use the current user of SharePoint and my delegated client id and use that to get a graph token? Invoking any login flow or consent permission request would work too.
Thanks for your time.

**Comment by andrewconnell** [comment]
Not directly with the SharePoint Framework APIs provided. The SPFx APIs use the SPO Azure AD app to get a token to call an endpoint like Microsoft Graph. In that case, the currently logged in user is used in combination with the SPO client ID.

If the file you're trying to download is in SPO, why use Graph? Just use the SharePoint REST API. That removes all extra access token & Microsoft Graph requirements. The current user's identity is used when accessing the SharePoint REST API from within a SPO site, plus the **SpHttpClient** API in the SPFx API handles the plumbing for you.

**Comment by RafatRashid** [comment]
Thats a good point. I am using SP rest apis for all sorts of file I/O. But I need a docx/xlsx/ppt file in pdf. I found <a href="https://docs.microsoft.com/en-us/graph/api/driveitem-get-content-format?view=graph-rest-1.0&tabs=http">this</a> Graph Api endpoint is the easiest and free way to get pdf of a sharepoint file.
That is why I am trying to access Graph API.

**Comment by andrewconnell** [comment]
@RafatRashid said
> the easiest and free way to get pdf of a sharepoint file

*Except* for your scenario which requires additional permissions to an API (Microsoft Graph)... and you have business process challenges. IMHO, the SP REST API will solve your problem with less friction for your situation.

**Comment by RafatRashid** [comment]
I dont believe SP rest is going to help me much with PDF conversion. I have been looking into MSAL.js to try to get access token for the signed in user.
So let me rephrase the base of my question. Is there any way I can use the identity of the SP signed in user to issue a token for any other AD protected resource (which in my case would be graph API)? I believe this falls under SPFX development issue?

Thank you very much @andrewconnell for your time.

**Comment by andrewconnell** [comment]
Yes... that's just implementing the AAD implicit flow. If the user is signed into SPO, they are signed into their AAD tenant.

@RafatRashid said:
> I believe this falls under SPFX development issue?

Not really... SPFx supports, through it's APIs, to call MS Graph & Azure AD secured endpoints. Those work by granting permission to the provided SPO AAD app (as discussed above). The included **HttpClient** can be used to call any endpoint, but you have to handle all the auth yourself.

Possible? Sure... but it's not a SharePoint issue. I recommend checking the Azure AD docs for implementing MSAL.js & the implicit flow for SPA apps (ie: 100% client-side apps) as that's as close to what you're looking for.
